pallets/interest-accrual/src/lib.rs

```diff
@@ -204,35 +205,56 @@ pub mod pallet {
 	#[pallet::genesis_build]
 	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
 		fn build(&self) {
-			StorageVersion::<T>::put(Release::V1);
+			StorageVersion::<T>::put(Release::V2);
 		}
 	}
 
 	#[pallet::hooks]
 	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
 		fn on_initialize(_: T::BlockNumber) -> Weight {
-			let mut count = 0;
 			let then = LastUpdated::<T>::get();
 			let now = Self::now();
 			LastUpdated::<T>::set(Self::now());
 			let delta = Self::now() - then;
 			let bits = Moment::BITS - delta.leading_zeros();
-			Rate::<T>::translate(|per_sec, mut rate: RateDetailsOf<T>| {
-				count += 1;
-				Self::calculate_accumulated_rate(per_sec, rate.accumulated_rate, then, now)
+
+			// reads: timestamp, last updated, rates vec
+			// writes: last updated, rates vec
+			let mut weight = T::DbWeight::get().reads_writes(3, 2);
+
+			let rates = Rates::<T>::get();
+			let rates: Vec<_> = rates
+				.into_iter()
+				.filter_map(|rate| {
+					weight = weight.saturating_add(T::Weights::calculate_accumulated_rate(bits));
```

Comment:
Could use `saturating_accrue`
```suggestion
					weight.saturating_accrue(T::Weights::calculate_accumulated_rate(bits));
```